[java.math] Added BigNums

Aliamondo opened the pull request:
Added the functionality of BigInteger, BigDecimal and MathContext to sdllib

Commit:
Added BigNums

Commit:
Fixed package name

Aliamondo commented:
Had wrong package names, because I renamed them by hand without an editor

budaimartin commented on src/hu/elte/txtuml/stdlib/math/MathContext.java:
```diff
@@ -0,0 +1,127 @@
+	 *         a {@code MathContext} object which has exactly the same
+	 *         settings as this object
+	 */
+	public boolean equals(Object x);
```
The `x` parameter should be `MathContext` instead of `Object` because `Object` is not a txtUML class nor an external class.

budaimartin commented on src/hu/elte/txtuml/stdlib/math/MathContextImplementation.java:
```diff
@@ -0,0 +1,106 @@
+	/*
+	 * Extract java.math.MathContext from MathContext
+	 */
+	protected static java.math.MathContext getOrigMathContext(Object val) {
```
What are you using this function for? Is it really needed?

budaimartin commented on src/hu/elte/txtuml/stdlib/math/BigInteger.java:
```diff
@@ -0,0 +1,633 @@
+	 * @return {@code true} if and only if the specified Object is a
+	 *         BigInteger whose value is numerically equal to this BigInteger.
+	 */
+	public boolean equals(Object x);
```
The parameter `x` should be type `BigInteger` instead of `Object` because `Object` is not a txtUML type nor an external class.

budaimartin commented on src/hu/elte/txtuml/stdlib/math/BigIntegerImplementation.java:
```diff
@@ -0,0 +1,398 @@
+	/*
+	 * Extract java.math.BigInteger from BigInteger
+	 */
+	protected static java.math.BigInteger getOrigBigInteger(Object val) {
```
What are you using this function for? Is it really needed?

budaimartin commented on src/hu/elte/txtuml/stdlib/math/BigIntegerImplementation.java:
```diff
@@ -0,0 +1,398 @@
+
+	@Override
+	public BigInteger add(BigInteger val) {
+		return new BigIntegerImplementation(num.add(getOrigBigInteger(val)));
```
Why don't you use `val.num` here and in the other functions below?

Aliamondo replied:
The problem is that `BigInteger` interface class doesn't have a `num` field, so I can't use `val.num`. For that to work in a way you specified, I'd need to change the method signature to
`public BigInteger add(BigIntegerImplementation val);`
Is this what you want?

budaimartin replied:
I see. You can either change the signature or cast `val` to `BigIntegerImplementation` in the parameter. It would be nice if we could get rid of the `getOrigBigInteger` function. Maybe @devaigergely81 can suggest other options.

budaimartin replied:
Nevermind. If you have discussed this with Gergely and Gábor then it will be fine.

budaimartin commented on src/hu/elte/txtuml/stdlib/math/BigDecimal.java:
```diff
@@ -0,0 +1,1308 @@
+	 * @see    #compareTo(BigDecimal)
+	 * @see    #hashCode
+	 */
+	public boolean equals(Object x);
```
The `obj` parameter should be type `UUID` instead of `Object` because `Object` is not a txtUML class nor an external class.

budaimartin commented on src/hu/elte/txtuml/stdlib/math/BigDecimalImplementation.java:
```diff
@@ -0,0 +1,737 @@
+	/*
+	 * Extract java.math.BigDecimal from BigDecimal
+	 */
+	protected static java.math.BigDecimal getOrigBigDecimal(Object val) {
```
What are you using this function for? Is it really needed?

budaimartin commented on src/hu/elte/txtuml/stdlib/math/BigDecimalImplementation.java:
```diff
@@ -0,0 +1,737 @@
+	 *            {@code BigDecimal}.
+	 */
+	public BigDecimalImplementation(BigInteger val) {
+		this.num = new java.math.BigDecimal(BigIntegerImplementation.getOrigBigInteger(val));
```
Why dont you use `val.num` here and in the other functions below?

budaimartin reviewed (changes requested):
I have checked your pull request. See my comments below.

Aliamondo committed:
Changed equals(Object) to respective classes and added more javadoc

Aliamondo commented:
I have modified `equals(Object obj)` methods to take their respective classes as arguments, i.e.
`equals(BigInteger x)` for `BigInteger` class.
I have also added the reasoning behind the `getOrig...` methods in the comments above them, but the basic idea is that the only way I can extract an original object out of an interface (`BigInteger`) is by casting it to a `BigIntegerImplementation`, but the method has the better readability and sometimes other objects use each other's extracting methods (like `BigDecimal` needs an original `BigInteger` in its constructor).
We had a discussion with Gergely and Gábor about this today and they have approved this method, but asked for a good explanation on why this method is needed in the comments, too.

Aliamondo committed:
Reorded the extract methods to make them faster
